blockchain/chain_sync/src/sync.rs

```diff
@@ -268,6 +216,68 @@ where
         }
     }
 
+    async fn handle_gossip_block(
+        block: GossipBlock,
+        source: Option<PeerId>,
+        network: SyncNetworkContext<DB>,
+        channel: &Sender<(PeerId, FullTipset)>,
+    ) {
+        let source = match source.clone() {
+            Some(source) => source,
+            None => {
+                warn!("Got a GossipBlock with no Source sender. This should not happen based on Filecoin's GossipSub options");
+                return;
+            }
+        };
+        info!(
+            "Received block over GossipSub: {} from {}",
+            block.header.epoch(),
+            source
+        );
+
+        // Get bls_messages in the store or over Bitswap
+        let bls_messages: Vec<_> = block
+            .bls_messages
+            .into_iter()
+            .map(|m| network.bitswap_get::<UnsignedMessage>(m))
+            .collect();
+
+        let bls_messages = match try_join_all(bls_messages).await {
+            Ok(msgs) => msgs,
+            Err(e) => {
+                warn!("Failed to get UnsignedMessage: {}", e);
+                return;
+            }
+        };
+
+        // Get secp_messages in the store or over Bitswap
+        let secp_messages: Vec<_> = block
+            .secpk_messages
+            .into_iter()
+            .map(|m| network.bitswap_get::<SignedMessage>(m))
+            .collect();
+
+        let secp_messages = match try_join_all(secp_messages).await {
+            Ok(msgs) => msgs,
+            Err(e) => {
+                warn!("Failed to get SignedMessage: {}", e);
+                return;
+            }
+        };
```

Comment:
These are currently done in parallel, so the first secp message request is sent only after _all_ bls messages are received. If you would like to leave for another PR, can you add a TODO so it's less likely it gets missed in the future?

Reply:
Sounds very reasonable, let me know if this was the fix that you had in mind here

Reply:
lgtm!